File: cpp/tests/unit/test_bugs_new24.cpp
// TDD failing tests for BUG-AUDIT-47 — CD T0SZ/T1SZ minimum bound check.
//
// ARM IHI0070G.b §5.4 CDTxSZInvalid():
//   For STT=0, VAX=0, IAS=48 (this model): txsz_min=16, txsz_max=39.
//   Mandatory check for SMMUv3.2+.
//
// Current code only checks the maximum (>39). Values in [1,15] are accepted
// silently — no C_BAD_CD event, no error return.
//
// BEFORE FIX: t0sz=10 accepted silently → tests FAIL.
// AFTER FIX:  C_BAD_CD emitted, error returned → tests PASS.

#include <gtest/gtest.h>
#include "smmu/smmu.h"

using namespace smmu;

// Helper: enable all queues and the SMMU global enable.
static void enableSMMU(SMMU& s) {
    s.setCR0(SMMU::CR0_SMMUEN | SMMU::CR0_EVENTQEN | SMMU::CR0_CMDQEN | SMMU::CR0_PRIQEN);
    s.setS1PSupported(true);
}

// Helper: check whether any event of the given type is present in the event queue.
static bool hasEvent(SMMU& smmu, EventType type) {
    std::vector<EventEntry> events = smmu.getEventQueue();
    for (const auto& ev : events) {
        if (ev.type == type) {
            return true;
        }
    }
    return false;
}

// Helper: configure and enable a stage-1 stream with the given t0sz/t1sz.
// Optionally maps a page at 0x1000 so translateUnlocked() is reached.
static void setupStage1Stream(SMMU& smmu, StreamID sid,
                               uint8_t t0sz, uint8_t t1sz = 0u,
                               bool mapPages = true) {
    StreamConfig cfg;
    cfg.translationEnabled = true;
    cfg.stage1Enabled      = true;
    cfg.stage2Enabled      = false;
    cfg.aa64               = true;
    cfg.t0sz               = t0sz;
    cfg.t1sz               = t1sz;
    cfg.faultMode          = FaultMode::Terminate;
    cfg.securityState      = SecurityState::NonSecure;
    cfg.vmid               = 0u;
    smmu.configureStream(sid, cfg);
    smmu.enableStream(sid);
    smmu.createStreamPASID(sid, 0u);
    if (mapPages) {
        PagePermissions perms(true, false, false);
        smmu.mapPage(sid, 0u, 0x1000u, 0x80001000u, perms);
    }
}

// ─── BUG-AUDIT-47 tests — must FAIL before fix, PASS after fix ───────────────

// Negative: t0sz=10 is below txsz_min=16.  translateUnlocked() must emit
// C_BAD_CD (ARM §5.4 CDTxSZInvalid).
//
// BEFORE FIX: only the >39 check exists → t0sz=10 is silently accepted →
//             no C_BAD_CD enqueued → test FAILS.
// AFTER FIX:  t0sz != 0 && t0sz < 16 triggers generateEvent(C_BAD_CD) → PASS.
TEST(BugAudit47, T0SzBelowMin_EmitsEvent) {
    SMMU smmu;
    enableSMMU(smmu);

    const StreamID sid  = 200u;
    const IOVA     iova = 0x1000u;

    setupStage1Stream(smmu, sid, /*t0sz=*/10u);

    smmu.translate(sid, 0u, iova, AccessType::Read, SecurityState::NonSecure);

    EXPECT_TRUE(hasEvent(smmu, EventType::C_BAD_CD))
        << "BUG-AUDIT-47: t0sz=10 < 16 must emit C_BAD_CD "
           "(ARM §5.4 CDTxSZInvalid: txsz_min=16 for STT=0, VAX=0)";
}

// Negative: t0sz=16 (valid), t1sz=10 (below minimum) → C_BAD_CD must be emitted.
// The spec check applies to both T0SZ and T1SZ independently.
//
// BEFORE FIX: only the >39 check exists for both fields → t1sz=10 silently
//             accepted → no event → test FAILS.
// AFTER FIX:  t1sz != 0 && t1sz < 16 triggers generateEvent(C_BAD_CD) → PASS.
TEST(BugAudit47, T1SzBelowMin_EmitsEvent) {
    SMMU smmu;
    enableSMMU(smmu);

    const StreamID sid  = 201u;
    const IOVA     iova = 0x1000u;

    setupStage1Stream(smmu, sid, /*t0sz=*/16u, /*t1sz=*/10u);

    smmu.translate(sid, 0u, iova, AccessType::Read, SecurityState::NonSecure);

    EXPECT_TRUE(hasEvent(smmu, EventType::C_BAD_CD))
        << "BUG-AUDIT-47: t1sz=10 < 16 must emit C_BAD_CD "
           "(ARM §5.4 CDTxSZInvalid: txsz_min applies to T1SZ as well)";
}

// Negative: t0sz=10 → translate() must return an error result (not ok).
// Exercises the return value path rather than the event queue.
//
// BEFORE FIX: no minimum-bound check → translateUnlocked() proceeds, reaches
//             VA range check (which passes because 2^(64-10) is huge), and
//             either finds the mapping or faults for a different reason.
//             In any case CDTxSZInvalid is not caught → result.isOk() may be
//             true → test FAILS.
// AFTER FIX:  makeTranslationError(SMMUError::InvalidConfiguration) returned
//             immediately → result.isOk() == false → test PASSES.
TEST(BugAudit47, T0SzBelowMin_TranslateFails) {
    SMMU smmu;
    enableSMMU(smmu);

    const StreamID sid  = 202u;
    const IOVA     iova = 0x1000u;

    setupStage1Stream(smmu, sid, /*t0sz=*/10u);

    TranslationResult result = smmu.translate(sid, 0u, iova,
                                              AccessType::Read,
                                              SecurityState::NonSecure);

    EXPECT_FALSE(result.isOk())
        << "BUG-AUDIT-47: t0sz=10 < 16 must cause translate() to return an error "
           "(ARM §5.4 CDTxSZInvalid → C_BAD_CD → InvalidConfiguration)";
}

// ─── Regression guard tests — must PASS both before AND after the fix ─────────

// Positive: t0sz=0 is the model sentinel for "no VA range restriction".
// The fix must NOT reject this value; the exclusion condition is (t0sz != 0u).
//
// This test passes before and after the fix.
TEST(BugAudit47, T0SzSentinelZero_Accepted) {
    SMMU smmu;
    enableSMMU(smmu);

    const StreamID sid  = 203u;
    const IOVA     iova = 0x1000u;

    setupStage1Stream(smmu, sid, /*t0sz=*/0u);

    TranslationResult result = smmu.translate(sid, 0u, iova,
                                              AccessType::Read,
                                              SecurityState::NonSecure);

    EXPECT_FALSE(hasEvent(smmu, EventType::C_BAD_CD))
        << "BUG-AUDIT-47: t0sz=0 (sentinel: no restriction) must NOT emit C_BAD_CD — "
           "the fix must only reject non-zero values below 16";
    EXPECT_TRUE(result.isOk())
        << "BUG-AUDIT-47: t0sz=0 (sentinel) must allow successful translation";
}

// Positive: t0sz=16 is exactly txsz_min — the lower boundary must be accepted.
// t0sz=16 → valid IOVAs are [0, 2^(64-16)) = [0, 2^48).  0x1000 is well within range.
//
// This test passes before and after the fix.
TEST(BugAudit47, T0SzAtMinBoundary_Accepted) {
    SMMU smmu;
    enableSMMU(smmu);

    const StreamID sid  = 204u;
    const IOVA     iova = 0x1000u;

    setupStage1Stream(smmu, sid, /*t0sz=*/16u);

    TranslationResult result = smmu.translate(sid, 0u, iova,
                                              AccessType::Read,
                                              SecurityState::NonSecure);

    EXPECT_FALSE(hasEvent(smmu, EventType::C_BAD_CD))
        << "BUG-AUDIT-47: t0sz=16 (lower boundary of valid range [16,39]) must NOT emit C_BAD_CD "
           "(ARM §5.4 CDTxSZInvalid: txsz_min=16)";
    EXPECT_TRUE(result.isOk())
        << "BUG-AUDIT-47: t0sz=16 (lower boundary) must allow successful translation";
}

// Positive: t0sz=39 is exactly txsz_max — the upper boundary must be accepted.
// t0sz=39 → valid IOVAs are [0, 2^(64-39)) = [0, 2^25) = [0, 32 MiB).
// IOVA=0x1000 is within that window.
// The existing >39 check already handled this correctly; this test guards against
// any regression introduced by the minimum-bound fix.
//
// This test passes before and after the fix.
TEST(BugAudit47, T0SzAtMaxBoundary_Accepted) {
    SMMU smmu;
    enableSMMU(smmu);

    const StreamID sid  = 205u;
    const IOVA     iova = 0x1000u;  // well within 2^25 = 32 MiB

    setupStage1Stream(smmu, sid, /*t0sz=*/39u);

    TranslationResult result = smmu.translate(sid, 0u, iova,
                                              AccessType::Read,
                                              SecurityState::NonSecure);

    EXPECT_FALSE(hasEvent(smmu, EventType::C_BAD_CD))
        << "BUG-AUDIT-47: t0sz=39 (upper boundary of valid range [16,39]) must NOT emit C_BAD_CD "
           "(ARM §5.4 CDTxSZInvalid: txsz_max=39)";
    EXPECT_TRUE(result.isOk())
        << "BUG-AUDIT-47: t0sz=39 (upper boundary) must allow successful translation";
}

File: rust/smmu/tests/test_bugs_new24.rs
//! TDD failing tests for BUG-AUDIT-47 (T0SZ/T1SZ min bound) and BUG-AUDIT-48 (s2_t0sz sentinel).
//!
//! BUG-AUDIT-47 (§5.4 CDTxSZInvalid): valid range [16,39] for this model. The minimum
//! bound (t0sz/t1sz < 16) is not currently checked. Values in [1,15] are accepted silently.
//!
//! BUG-AUDIT-48: configure_stream() rejects s2_t0sz=0 (sentinel) — the BUG-AUDIT-45 fix
//! omitted the `!= 0` guard that C++ correctly includes. Stage-2 streams with s2_t0sz=0
//! should be accepted (0 means "no IPA range restriction").
#![allow(missing_docs)]
#![allow(clippy::unwrap_used)]
#![allow(clippy::items_after_statements)]
#![allow(clippy::doc_markdown)]
#![allow(clippy::uninlined_format_args)]

use smmu::types::{
    AccessType, EventType, PagePermissions, SecurityState, StreamConfig, StreamID, IOVA, PA, PASID,
};
use smmu::SMMU;

// ============================================================================
// Helpers
// ============================================================================

fn sid(n: u32) -> StreamID {
    StreamID::new(n).unwrap()
}
fn pasid0() -> PASID {
    PASID::new(0).unwrap()
}
fn iova(addr: u64) -> IOVA {
    IOVA::new(addr).unwrap()
}
fn pa(addr: u64) -> PA {
    PA::new(addr).unwrap()
}

/// Build an SMMU with all queues and global enable active.
fn make_smmu() -> SMMU {
    let smmu = SMMU::new();
    smmu.set_cr0(SMMU::CR0_SMMUEN | SMMU::CR0_CMDQEN | SMMU::CR0_EVENTQEN | SMMU::CR0_PRIQEN);
    smmu
}

/// Returns true when the event queue contains at least one event of the given type.
fn has_event(smmu: &SMMU, event_type: EventType) -> bool {
    smmu.get_events().iter().any(|e| e.event_type == event_type)
}

// ============================================================================
// BUG-AUDIT-47: T0SZ/T1SZ below minimum bound [16] must emit C_BAD_CD
// ============================================================================

/// BUG-AUDIT-47 test 1: t0sz=10 (below minimum 16) must emit C_BAD_CD event on translate.
///
/// ARM §5.4 CDTxSZInvalid(): for VAX=0, STT=0, IAS=48, valid T0SZ range is [16, 39].
/// Values in [1,15] are silently accepted by the current code — no C_BAD_CD event is
/// generated and the translate call does not return Err.
///
/// BEFORE FIX: translate() returns Ok (no event emitted) → FAILS.
/// AFTER FIX:  translate() returns Err AND CBadCd event emitted → PASSES.
#[test]
fn t0sz_below_min_emits_event() {
    let smmu = make_smmu();
    smmu.set_s1p_supported(true);

    let mut cfg = StreamConfig::stage1_only();
    cfg.security_state = SecurityState::NonSecure;
    cfg.t0sz = 10; // below minimum of 16
    cfg.t1sz = 0; // sentinel — no restriction
    smmu.configure_stream(sid(200), cfg).unwrap();
    smmu.create_pasid(sid(200), pasid0()).unwrap();
    smmu.map_page(
        sid(200),
        pasid0(),
        iova(0x1000),
        pa(0xA000),
        PagePermissions::read_only(),
        SecurityState::NonSecure,
    )
    .unwrap();

    let _ = smmu.translate(
        sid(200),
        pasid0(),
        iova(0x1000),
        AccessType::Read,
        SecurityState::NonSecure,
    );

    assert!(
        has_event(&smmu, EventType::CBadCd),
        "BUG-AUDIT-47: C_BAD_CD event must be emitted when t0sz=10 < 16 \
         (ARM §5.4 CDTxSZInvalid — minimum bound not currently checked)"
    );
}

/// BUG-AUDIT-47 test 2: t1sz=10 (below minimum 16) must emit C_BAD_CD event on translate.
///
/// ARM §5.4 CDTxSZInvalid(): for VAX=0, STT=0, IAS=48, valid T1SZ range is [16, 39].
/// When t0sz is valid (16) but t1sz is below the minimum, C_BAD_CD must still be emitted.
///
/// BEFORE FIX: translate() returns Ok (no event emitted) → FAILS.
/// AFTER FIX:  translate() returns Err AND CBadCd event emitted → PASSES.
#[test]
fn t1sz_below_min_emits_event() {
    let smmu = make_smmu();
    smmu.set_s1p_supported(true);

    let mut cfg = StreamConfig::stage1_only();
    cfg.security_state = SecurityState::NonSecure;
    cfg.t0sz = 16; // valid
    cfg.t1sz = 10; // below minimum of 16
    smmu.configure_stream(sid(201), cfg).unwrap();
    smmu.create_pasid(sid(201), pasid0()).unwrap();
    smmu.map_page(
        sid(201),
        pasid0(),
        iova(0x1000),
        pa(0xB000),
        PagePermissions::read_only(),
        SecurityState::NonSecure,
    )
    .unwrap();

    let _ = smmu.translate(
        sid(201),
        pasid0(),
        iova(0x1000),
        AccessType::Read,
        SecurityState::NonSecure,
    );

    assert!(
        has_event(&smmu, EventType::CBadCd),
        "BUG-AUDIT-47: C_BAD_CD event must be emitted when t1sz=10 < 16 \
         (ARM §5.4 CDTxSZInvalid — t1sz minimum bound not currently checked)"
    );
}

/// BUG-AUDIT-47 test 3: t0sz=10 (below minimum 16) must cause translate to return Err.
///
/// ARM §5.4 CDTxSZInvalid(): out-of-range T0SZ values must abort translation with C_BAD_CD.
///
/// BEFORE FIX: translate() returns Ok (silently proceeds) → FAILS.
/// AFTER FIX:  translate() returns Err → PASSES.
#[test]
fn t0sz_below_min_translate_fails() {
    let smmu = make_smmu();
    smmu.set_s1p_supported(true);

    let mut cfg = StreamConfig::stage1_only();
    cfg.security_state = SecurityState::NonSecure;
    cfg.t0sz = 10; // below minimum of 16
    cfg.t1sz = 0; // sentinel — no restriction
    smmu.configure_stream(sid(202), cfg).unwrap();
    smmu.create_pasid(sid(202), pasid0()).unwrap();
    smmu.map_page(
        sid(202),
        pasid0(),
        iova(0x1000),
        pa(0xC000),
        PagePermissions::read_only(),
        SecurityState::NonSecure,
    )
    .unwrap();

    let result = smmu.translate(
        sid(202),
        pasid0(),
        iova(0x1000),
        AccessType::Read,
        SecurityState::NonSecure,
    );

    assert!(
        result.is_err(),
        "BUG-AUDIT-47: translate() must return Err when t0sz=10 < 16 \
         (ARM §5.4 CDTxSZInvalid — translation must be aborted for out-of-range T0SZ)"
    );
}

/// BUG-AUDIT-47 test 4 (regression): t0sz=0 (sentinel, no restriction) must be accepted.
///
/// The model uses t0sz=0 throughout as a "no restriction" sentinel. This must NOT be
/// treated as below-minimum by the new minimum-bound check.
///
/// BEFORE / AFTER FIX: This must always pass (regression guard).
#[test]
fn t0sz_sentinel_zero_accepted() {
    let smmu = make_smmu();
    smmu.set_s1p_supported(true);

    let mut cfg = StreamConfig::stage1_only();
    cfg.security_state = SecurityState::NonSecure;
    cfg.t0sz = 0; // sentinel — no VA-range restriction
    cfg.t1sz = 0; // sentinel — no restriction
    smmu.configure_stream(sid(203), cfg).unwrap();
    smmu.create_pasid(sid(203), pasid0()).unwrap();
    smmu.map_page(
        sid(203),
        pasid0(),
        iova(0x1000),
        pa(0xD000),
        PagePermissions::read_only(),
        SecurityState::NonSecure,
    )
    .unwrap();

    let result = smmu.translate(
        sid(203),
        pasid0(),
        iova(0x1000),
        AccessType::Read,
        SecurityState::NonSecure,
    );

    assert!(
        result.is_ok(),
        "BUG-AUDIT-47: translate() must succeed when t0sz=0 (sentinel, no VA restriction); \
         0 must NOT be rejected by the minimum-bound check (regression guard)"
    );
    assert!(
        !has_event(&smmu, EventType::CBadCd),
        "BUG-AUDIT-47: no C_BAD_CD must be emitted for t0sz=0 (sentinel — no restriction)"
    );
}

/// BUG-AUDIT-47 test 5 (regression): t0sz=16 (minimum boundary) must be accepted.
///
/// t0sz=16 is the lower bound of the valid range [16, 39] and must not be rejected.
///
/// BEFORE / AFTER FIX: This must always pass (boundary regression guard).
#[test]
fn t0sz_at_min_boundary_accepted() {
    let smmu = make_smmu();
    smmu.set_s1p_supported(true);

    let mut cfg = StreamConfig::stage1_only();
    cfg.security_state = SecurityState::NonSecure;
    cfg.t0sz = 16; // minimum boundary — legal
    cfg.t1sz = 0;  // sentinel — no restriction
    smmu.configure_stream(sid(204), cfg).unwrap();
    smmu.create_pasid(sid(204), pasid0()).unwrap();
    smmu.map_page(
        sid(204),
        pasid0(),
        iova(0x1000),
        pa(0xE000),
        PagePermissions::read_only(),
        SecurityState::NonSecure,
    )
    .unwrap();

    let result = smmu.translate(
        sid(204),
        pasid0(),
        iova(0x1000),
        AccessType::Read,
        SecurityState::NonSecure,
    );

    assert!(
        result.is_ok(),
        "BUG-AUDIT-47: translate() must succeed for t0sz=16 (minimum boundary, ARM §5.4 CDTxSZInvalid)"
    );
    assert!(
        !has_event(&smmu, EventType::CBadCd),
        "BUG-AUDIT-47: no C_BAD_CD must be emitted for t0sz=16 (lower boundary)"
    );
}

/// BUG-AUDIT-47 test 6 (regression): t0sz=39 (maximum boundary) must be accepted.
///
/// t0sz=39 is the upper bound of the valid range [16, 39]. For t0sz=39, the valid
/// IOVA range is [0, 2^(64-39)) = [0, 2^25). Using IOVA=0x1000 (< 2^25) is valid.
///
/// BEFORE / AFTER FIX: This must always pass (boundary regression guard).
#[test]
fn t0sz_at_max_boundary_accepted() {
    let smmu = make_smmu();
    smmu.set_s1p_supported(true);

    let mut cfg = StreamConfig::stage1_only();
    cfg.security_state = SecurityState::NonSecure;
    cfg.t0sz = 39; // maximum boundary — legal; valid range [0, 2^25)
    cfg.t1sz = 0;  // sentinel — no restriction
    smmu.configure_stream(sid(205), cfg).unwrap();
    smmu.create_pasid(sid(205), pasid0()).unwrap();
    smmu.map_page(
        sid(205),
        pasid0(),
        iova(0x1000), // 0x1000 < 2^25 — within valid range for t0sz=39
        pa(0xF000),
        PagePermissions::read_only(),
        SecurityState::NonSecure,
    )
    .unwrap();

    let result = smmu.translate(
        sid(205),
        pasid0(),
        iova(0x1000),
        AccessType::Read,
        SecurityState::NonSecure,
    );

    assert!(
        result.is_ok(),
        "BUG-AUDIT-47: translate() must succeed for t0sz=39 (maximum boundary, ARM §5.4 CDTxSZInvalid)"
    );
    assert!(
        !has_event(&smmu, EventType::CBadCd),
        "BUG-AUDIT-47: no C_BAD_CD must be emitted for t0sz=39 (upper boundary)"
    );
}

// ============================================================================
// BUG-AUDIT-48: configure_stream() must accept s2_t0sz=0 (sentinel)
// ============================================================================

/// BUG-AUDIT-48 test 1: configure_stream() with s2_t0sz=0 (sentinel) must return Ok.
///
/// ARM §5.2: s2_t0sz=0 is the model's "no IPA range restriction" sentinel, identical to
/// C++ behavior. The BUG-AUDIT-45 fix introduced a pre-check at configure_stream() that
/// correctly rejects s2_t0sz in [1,15] and [40,255], but omitted the `!= 0` guard,
/// causing s2_t0sz=0 to be incorrectly rejected as "< 16".
///
/// BEFORE FIX: configure_stream() returns Err for s2_t0sz=0 → FAILS.
/// AFTER FIX:  configure_stream() returns Ok → PASSES.
#[test]
fn configure_stream_s2_t0sz_zero_sentinel_accepted() {
    let smmu = make_smmu();
    smmu.set_s2p_supported(true);

    let mut cfg = StreamConfig::stage2_only();
    cfg.security_state = SecurityState::NonSecure;
    cfg.s2_t0sz = 0; // sentinel — no IPA range restriction

    let result = smmu.configure_stream(sid(210), cfg);
    assert!(
        result.is_ok(),
        "BUG-AUDIT-48: configure_stream() must accept s2_t0sz=0 (sentinel = no IPA range \
         restriction); BUG-AUDIT-45 fix incorrectly rejects 0 as < 16 (ARM §5.2)"
    );
}

/// BUG-AUDIT-48 test 2: configure_stream() with s2_t0sz=0 must not emit C_BAD_STE event.
///
/// ARM §5.2: s2_t0sz=0 is a valid sentinel value. No C_BAD_STE event should be emitted.
///
/// BEFORE FIX: CBadSte event emitted → FAILS.
/// AFTER FIX:  no CBadSte event → PASSES.
#[test]
fn configure_stream_s2_t0sz_zero_no_bad_ste_event() {
    let smmu = make_smmu();
    smmu.set_s2p_supported(true);

    let mut cfg = StreamConfig::stage2_only();
    cfg.security_state = SecurityState::NonSecure;
    cfg.s2_t0sz = 0; // sentinel — no IPA range restriction

    let _ = smmu.configure_stream(sid(211), cfg);

    assert!(
        !has_event(&smmu, EventType::CBadSte),
        "BUG-AUDIT-48: no C_BAD_STE event must be emitted for s2_t0sz=0 (sentinel); \
         the BUG-AUDIT-45 pre-check incorrectly treats 0 as out-of-range (ARM §5.2)"
    );
}

/// BUG-AUDIT-48 test 3 (regression): configure_stream() with s2_t0sz=16 must still be accepted.
///
/// Valid value at the minimum boundary; must remain accepted both before and after the fix.
///
/// BEFORE / AFTER FIX: This must always pass (regression guard).
#[test]
fn configure_stream_s2_t0sz_valid_regression() {
    let smmu = make_smmu();
    smmu.set_s2p_supported(true);

    let mut cfg = StreamConfig::stage2_only();
    cfg.security_state = SecurityState::NonSecure;
    cfg.s2_t0sz = 16; // valid minimum boundary

    let result = smmu.configure_stream(sid(212), cfg);
    assert!(
        result.is_ok(),
        "BUG-AUDIT-48: configure_stream() must accept s2_t0sz=16 (valid minimum boundary, \
         ARM §5.2 — regression guard)"
    );
    assert!(
        !has_event(&smmu, EventType::CBadSte),
        "BUG-AUDIT-48: no C_BAD_STE must be emitted for valid s2_t0sz=16 (regression guard)"
    );
}
